lib/fs.js

```diff
@@ -2598,3 +2596,465 @@ Object.defineProperty(fs, 'SyncWriteStream', {
   set: internalUtil.deprecate((val) => { SyncWriteStream = val; },
                               'fs.SyncWriteStream is deprecated.', 'DEP0061')
 });
+
+// Promises API
+function validateFileHandle(handle) {
+  if (!(handle instanceof FileHandle))
+    throw new errors.TypeError('ERR_INVALID_ARG_TYPE',
+                               'filehandle', 'FileHandle');
+}
+
+
+async function writeFileHandle(filehandle, data, options) {
+  let buffer = isUint8Array(data) ?
+    data : Buffer.from('' + data, options.encoding || 'utf8');
+  let remaining = buffer.length;
+  if (remaining === 0) return;
+  do {
+    const { bytesWritten } =
+      await fs.promises.write(filehandle, buffer, 0,
+                              Math.min(16384, buffer.length));
+    remaining -= bytesWritten;
+    buffer = buffer.slice(bytesWritten);
+  } while (remaining > 0);
+}
+
+async function readFileHandle(filehandle, options) {
+  const req = new FSReqPromise();
+  binding.fstat(filehandle.fd, req);
+  await req.promise;
+
+  let size;
+  if ((req.statFields[1/*mode*/] & S_IFMT) === S_IFREG) {
+    size = req.statFields[8/*size*/];
+  } else {
+    size = 0;
+  }
+
+  if (size === 0)
+    return Buffer.alloc(0);
+
+  if (size > kMaxLength)
+    throw new errors.RangeError('ERR_BUFFER_TOO_LARGE');
+
+  const chunks = [];
+  const chunkSize = Math.min(size, 16384);
+  const buf = Buffer.alloc(chunkSize);
+  let read = 0;
+  do {
+    const { bytesRead, buffer } =
+      await fs.promises.read(filehandle, buf, 0, size);
+    read = bytesRead;
+    if (read > 0)
+      chunks.push(buffer.slice(0, read));
+  } while (read === chunkSize);
+
+  return Buffer.concat(chunks);
+}
+
+fs.promises = {
+  async access(path, mode = fs.F_OK) {
```

Comment:
Given that you never `await` anything, does this need to be `async`?

Reply:
To catch errors thrown, yes.

Reply:
@jasnell, can you elaborate? For my education, how does making the function async help with catching thrown errors? My understanding is that `async` merely enables the use of the `await` keyword.

Reply:
Any error thrown within an `async function` is automatically caught and causes rejection of the promise returned by the function. With a regular function you could have two cases: synchronous throw or rejection depending on where the error happens.

Reply:
Yeah, what @targos said... for instance, try the following cases:

```js
function foo() {
  throw new Error('foo');  // error is thrown synchronously
  return new Promise(() => {});
}

async function bar() {
  throw new Error('bar'); // error causes promise rejection
  return new Promise(() => {});
}
```